<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Manhattan plot — fig_manhattan • geni.plots</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Manhattan plot — fig_manhattan"><meta name="description" content="fig_manhattan creates a Manhattan plot for genomic markers
from across the genome, e.g. results from genome-wide association studies."><meta property="og:description" content="fig_manhattan creates a Manhattan plot for genomic markers
from across the genome, e.g. results from genome-wide association studies."><meta property="og:image" content="https://jrs95.github.io/geni.plots/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">geni.plots</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/geni_plots.html">GENI plots</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jrs95/geni.plots/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Manhattan plot</h1>
      <small class="dont-index">Source: <a href="https://github.com/jrs95/geni.plots/blob/v0.1.2/R/fig_manhattan.R" class="external-link"><code>R/fig_manhattan.R</code></a></small>
      <div class="d-none name"><code>fig_manhattan.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>fig_manhattan</code> creates a Manhattan plot for genomic markers
from across the genome, e.g. results from genome-wide association studies.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fig_manhattan</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#7395D3"</span>, <span class="st">"#1D345D"</span><span class="op">)</span>,</span>
<span>  rank_pos <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  thin_thresh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  block_thresh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  interactive_n <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thresh <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e-05</span>, <span class="fl">5e-08</span><span class="op">)</span>,</span>
<span>  thresh_size <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  thresh_colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey50"</span>, <span class="st">"red"</span><span class="op">)</span>,</span>
<span>  trunc <span class="op">=</span> <span class="fl">1e-30</span>,</span>
<span>  highlight_colours <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  point_size <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  chr_dist <span class="op">=</span> <span class="fl">1e+07</span>,</span>
<span>  x_labels <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_top <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  label_thresh <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  label_ylim <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">1e-05</span><span class="op">)</span>,</span>
<span>  label_nudge_y <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  label_box <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  title_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  title_center <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  axis_text_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  axis_title_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_width <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  plot_height <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  girafe <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>a <code>data.frame</code> containing the chromosome-position information and
association statistics for each genomic marker (e.g., genetic variants)
with the following columns:</p><ul><li><p><code>chr</code>
the chromosome for each genomic marker</p></li>
<li><p><code>pos</code>
the genomic position for each genomic marker</p></li>
<li><p><code>pvalue</code>
the association p-value for each genomic marker</p></li>
<li><p><code>highlight</code>
the optional highlight points variable, where
<code>0</code> = point not highlighted,
<code>1</code> = first highlight colour,
<code>2</code> = second highlight colour, etc.</p></li>
<li><p><code>highlight_shape</code>
the optional highlight point shape variable, where
<code>0</code> = standard circle,
<code>1</code> = standard circle with border,
<code>2</code> = standard rectangle,
<code>3</code> = standard rectangle with border,
<code>4</code> = standard diamond,
<code>5</code> = standard diamond with border</p></li>
<li><p><code>label</code>
the optional point labelling variable (e.g. gene name),
if <code>label = ""</code> for a point then no
label is presented for that point</p></li>
<li><p><code>text</code>
the optional hover text variable for interactive plots to
display further information,
if <code>text = ""</code> for a point then no hover text is
presented for that point</p></li>
</ul></dd>


<dt id="arg-colours">colours<a class="anchor" aria-label="anchor" href="#arg-colours"></a></dt>
<dd><p>a <code>vector</code> of colours used to differentiate chromosomes,
these can either be a separate colour for each chromosome or
a pair of colours (default: <code>c("#7395D3", "#1D345D")</code>)</p></dd>


<dt id="arg-rank-pos">rank_pos<a class="anchor" aria-label="anchor" href="#arg-rank-pos"></a></dt>
<dd><p>a <code>logical</code> value whether genomic markers should be
plotted by their rank position (default: <code>FALSE</code>)</p></dd>


<dt id="arg-thin-thresh">thin_thresh<a class="anchor" aria-label="anchor" href="#arg-thin-thresh"></a></dt>
<dd><p>a <code>numeric</code> value representing the minimum p-value
threshold for genomic markers to be displayed (default: <code>NULL</code>)</p></dd>


<dt id="arg-block-thresh">block_thresh<a class="anchor" aria-label="anchor" href="#arg-block-thresh"></a></dt>
<dd><p>a <code>numeric</code> value for representing a p-value threshold,
above which genomic markers are represented using blocks of colour
(default: <code>NULL</code>)</p></dd>


<dt id="arg-interactive">interactive<a class="anchor" aria-label="anchor" href="#arg-interactive"></a></dt>
<dd><p>a <code>logical</code> value indicating whether the plot
should be interactive (default: <code>FALSE</code>)</p></dd>


<dt id="arg-interactive-n">interactive_n<a class="anchor" aria-label="anchor" href="#arg-interactive-n"></a></dt>
<dd><p>a <code>numeric</code> value indicating the number of top
associated points (max = 100,000) to present before using blocks of colour
to minimise file size (default: <code>NULL</code>)</p></dd>


<dt id="arg-thresh">thresh<a class="anchor" aria-label="anchor" href="#arg-thresh"></a></dt>
<dd><p>a <code>numeric</code> <code>vector</code> providing p-value thresholds to be plotted
(default: <code>c(1e-5, 5e-8)</code>)</p></dd>


<dt id="arg-thresh-size">thresh_size<a class="anchor" aria-label="anchor" href="#arg-thresh-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the width of the lines
indicating the p-value thresholds (default: <code>0.5</code>)</p></dd>


<dt id="arg-thresh-colours">thresh_colours<a class="anchor" aria-label="anchor" href="#arg-thresh-colours"></a></dt>
<dd><p>a <code>character</code> <code>vector</code> indicating the colours of the
lines indicating the p-value thresholds (default: <code>c("grey50", "red")</code>)</p></dd>


<dt id="arg-trunc">trunc<a class="anchor" aria-label="anchor" href="#arg-trunc"></a></dt>
<dd><p>a <code>numeric</code> value representing the maximum p-value for which
results are displayed (default: <code>1e-30</code>)</p></dd>


<dt id="arg-highlight-colours">highlight_colours<a class="anchor" aria-label="anchor" href="#arg-highlight-colours"></a></dt>
<dd><p>a <code>vector</code> indicating the colours for the
highlighted genomic markers (default: <code>NULL</code>)</p></dd>


<dt id="arg-point-size">point_size<a class="anchor" aria-label="anchor" href="#arg-point-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of each point
(default: <code>2</code>)</p></dd>


<dt id="arg-chr-dist">chr_dist<a class="anchor" aria-label="anchor" href="#arg-chr-dist"></a></dt>
<dd><p>a <code>numeric</code> value indicating the gap between different
chromosomes (default: <code>10000000</code>)</p></dd>


<dt id="arg-x-labels">x_labels<a class="anchor" aria-label="anchor" href="#arg-x-labels"></a></dt>
<dd><p>a <code>logical</code> value whether the x-axis should be labelled
(default: <code>TRUE</code>)</p></dd>


<dt id="arg-label-top">label_top<a class="anchor" aria-label="anchor" href="#arg-label-top"></a></dt>
<dd><p>a <code>logical</code> value whether the top associated points should
be labelled (default: <code>TRUE</code>)</p></dd>


<dt id="arg-label-thresh">label_thresh<a class="anchor" aria-label="anchor" href="#arg-label-thresh"></a></dt>
<dd><p>a <code>numeric</code> value providing a p-value threshold for
labelling points (default: <code>1e-5</code>)</p></dd>


<dt id="arg-label-size">label_size<a class="anchor" aria-label="anchor" href="#arg-label-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of each label
(default: <code>3</code>)</p></dd>


<dt id="arg-label-ylim">label_ylim<a class="anchor" aria-label="anchor" href="#arg-label-ylim"></a></dt>
<dd><p>a <code>numeric</code> value indicating maximum y-axis value at which
labels can be displayed (default: <code>-log10(1e-5)</code>)</p></dd>


<dt id="arg-label-nudge-y">label_nudge_y<a class="anchor" aria-label="anchor" href="#arg-label-nudge-y"></a></dt>
<dd><p>a <code>numeric</code> value indicating the degree to which label
placement on the y-axis should be adjusted (default: <code>0</code>)`</p></dd>


<dt id="arg-label-box">label_box<a class="anchor" aria-label="anchor" href="#arg-label-box"></a></dt>
<dd><p>a <code>logical</code> value indicating whether labels should be
surrounded by a box (default: <code>FALSE</code>)</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>a <code>character`` string providing a title for the plot (default: </code>NULL`)</p></dd>


<dt id="arg-title-size">title_size<a class="anchor" aria-label="anchor" href="#arg-title-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of the title text
for the plot (default: <code>NULL</code>)</p></dd>


<dt id="arg-title-center">title_center<a class="anchor" aria-label="anchor" href="#arg-title-center"></a></dt>
<dd><p>a <code>numeric</code> value indicating whether the plot title
should be centered (default: <code>FALSE</code>)</p></dd>


<dt id="arg-axis-text-size">axis_text_size<a class="anchor" aria-label="anchor" href="#arg-axis-text-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of the axis
text for the plot (default: <code>NULL</code>)</p></dd>


<dt id="arg-axis-title-size">axis_title_size<a class="anchor" aria-label="anchor" href="#arg-axis-title-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of the axis
title text for the plot (default: <code>NULL</code>)</p></dd>


<dt id="arg-plot-width">plot_width<a class="anchor" aria-label="anchor" href="#arg-plot-width"></a></dt>
<dd><p>a <code>numeric</code> value indicating the width of the plot
(default: <code>16</code>)</p></dd>


<dt id="arg-plot-height">plot_height<a class="anchor" aria-label="anchor" href="#arg-plot-height"></a></dt>
<dd><p>a <code>numeric</code> value indicating the height of the plot
(default: <code>8</code>)</p></dd>


<dt id="arg-girafe">girafe<a class="anchor" aria-label="anchor" href="#arg-girafe"></a></dt>
<dd><p>a <code>logical</code> value indicating whether an interactive plot
should be turned into an interactive graphic using
<a href="https://davidgohel.github.io/ggiraph/reference/girafe.html" class="external-link">girafe()</a> (default = <code>TRUE</code>)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>fig_manhattan</code> returns a Manhattan plot for genomic markers
from across the genome, e.g. results from genome-wide association studies.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This plotting function plots a Manhattan plot for genomic markers
from across the genome. The default is to truncate these results to
p-value cut-off of <code>1e-30</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>James Staley <a href="mailto:jrstaley95@gmail.com">jrstaley95@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">fig_manhattan</span><span class="op">(</span></span></span>
<span class="r-in"><span>  data <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="geni_test_manhattan.html">geni_test_manhattan</a></span>,</span></span>
<span class="r-in"><span>  block_thresh <span class="op">=</span> <span class="fl">1e-4</span>,</span></span>
<span class="r-in"><span>  label_box <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="fig_manhattan-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Staley, Wes Spiller.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

