<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Regional plot — fig_region • geni.plots</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Regional plot — fig_region"><meta name="description" content="fig_region creates a regional plot, i.e. a scatter graph of
genomic markers associations (e.g. log10(p-values)) with a gene bar
underneath."><meta property="og:description" content="fig_region creates a regional plot, i.e. a scatter graph of
genomic markers associations (e.g. log10(p-values)) with a gene bar
underneath."><meta property="og:image" content="https://jrs95.github.io/geni.plots/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">geni.plots</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/geni_plots.html">GENI plots</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/jrs95/geni.plots/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Regional plot</h1>
      <small class="dont-index">Source: <a href="https://github.com/jrs95/geni.plots/blob/main/R/fig_region.R" class="external-link"><code>R/fig_region.R</code></a></small>
      <div class="d-none name"><code>fig_region.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>fig_region</code> creates a regional plot, i.e. a scatter graph of
genomic markers associations (e.g. log10(p-values)) with a gene bar
underneath.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fig_region</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  corr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  corr_top <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top_marker <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  r2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  build <span class="op">=</span> <span class="fl">38</span>,</span>
<span>  prob <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  thresh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thresh_colour <span class="op">=</span> <span class="st">"grey50"</span>,</span>
<span>  x_min <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  x_max <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  y_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  point_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  genebar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  genebar_ntracks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genebar_label_pos <span class="op">=</span> <span class="fl">3.6</span>,</span>
<span>  genebar_label_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  genebar_line_size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">3.5</span>,</span>
<span>  highlights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlights_cat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlights_label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  highlights_shape <span class="op">=</span> <span class="fl">22</span>,</span>
<span>  highlights_nolabel_shape <span class="op">=</span> <span class="fl">21</span>,</span>
<span>  highlights_sort <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  highlights_colours <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlights_title <span class="op">=</span> <span class="st">"Group"</span>,</span>
<span>  title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  title_size <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  title_center <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  axis_text_size <span class="op">=</span> <span class="fl">14</span>,</span>
<span>  axis_title_size <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  legend_text_size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  legend_title_size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  point_padding <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  nudge_x <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  nudge_y <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  nudge_y_top <span class="op">=</span> <span class="fl">0.06</span>,</span>
<span>  ylim_prob <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  assoc_plot_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genebar_plot_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_plot_dist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_width <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  plot_height <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  girafe <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data">data<a class="anchor" aria-label="anchor" href="#arg-data"></a></dt>
<dd><p>a <code>data.frame</code> containing the association statistics for each
marker within a genomic region with the following columns:</p><ul><li><p><code>marker</code>
the genomic marker identifier (e.g. rsID)</p></li>
<li><p><code>chr</code>
the chromosome for each genomic marker</p></li>
<li><p><code>pos</code>
the genomic position for each genomic marker

and one of the following:</p></li>
<li><p><code>pvalue</code>
the association p-value for each genomic marker</p></li>
<li><p><code>z</code>
the association z-statistic for each genomic marker</p></li>
<li><p><code>prob</code>
the association probability for each genomic marker</p></li>
</ul></dd>


<dt id="arg-corr">corr<a class="anchor" aria-label="anchor" href="#arg-corr"></a></dt>
<dd><p>a <code>numeric</code> <code>matrix</code> of correlation statistics between
the markers (default: <code>NULL</code>)</p></dd>


<dt id="arg-corr-top">corr_top<a class="anchor" aria-label="anchor" href="#arg-corr-top"></a></dt>
<dd><p>a <code>numeric</code> <code>vector</code> of correlation statistics between the
top marker and the rest of the markers (default: <code>NULL</code>)</p></dd>


<dt id="arg-top-marker">top_marker<a class="anchor" aria-label="anchor" href="#arg-top-marker"></a></dt>
<dd><p>a <code>character</code> value depicting the marker to plot the
correlation statistics of the rest of the markers against (default: <code>NULL</code>)</p></dd>


<dt id="arg-r-">r2<a class="anchor" aria-label="anchor" href="#arg-r-"></a></dt>
<dd><p>a <code>logical</code> value indicating whether the set of correlation
statistics entered in <code>corr</code> or <code>corr_top</code> are squared (default: <code>FALSE</code>)</p></dd>


<dt id="arg-build">build<a class="anchor" aria-label="anchor" href="#arg-build"></a></dt>
<dd><p>a <code>numeric</code> value indicating the genome build used to determine
genomic position (default: <code>38</code> representing human assembly GRCh38)</p></dd>


<dt id="arg-prob">prob<a class="anchor" aria-label="anchor" href="#arg-prob"></a></dt>
<dd><p>a <code>logical</code> value indicating whether probability statistics
should be plotted instead of -log10(p-values) (default: <code>FALSE</code>)</p></dd>


<dt id="arg-interactive">interactive<a class="anchor" aria-label="anchor" href="#arg-interactive"></a></dt>
<dd><p>a <code>logical</code> value indicating whether the plot
should be interactive (default: <code>FALSE</code>)</p></dd>


<dt id="arg-thresh">thresh<a class="anchor" aria-label="anchor" href="#arg-thresh"></a></dt>
<dd><p>a <code>numeric</code> <code>vector</code> providing the p-value thresholds to be
plotted (default: <code>NULL</code>)</p></dd>


<dt id="arg-thresh-colour">thresh_colour<a class="anchor" aria-label="anchor" href="#arg-thresh-colour"></a></dt>
<dd><p>a <code>character</code> <code>vector</code> indicating the colours of the
lines indicating the p-value thresholds (default: <code>"grey50"</code>)</p></dd>


<dt id="arg-x-min">x_min<a class="anchor" aria-label="anchor" href="#arg-x-min"></a></dt>
<dd><p>a <code>numeric</code> value depicting the minimum plotted x-axis value
representing the start of the genomic region (default: <code>NULL</code>)</p></dd>


<dt id="arg-x-max">x_max<a class="anchor" aria-label="anchor" href="#arg-x-max"></a></dt>
<dd><p>a <code>numeric</code> value depicting the maximum plotted x-axis value
representing the end of the genomic region (default: <code>NULL</code>)</p></dd>


<dt id="arg-y-title">y_title<a class="anchor" aria-label="anchor" href="#arg-y-title"></a></dt>
<dd><p>a <code>character</code> string defining the title of the y-axis
(default: <code>NULL</code>)</p></dd>


<dt id="arg-point-size">point_size<a class="anchor" aria-label="anchor" href="#arg-point-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of each point
(default: <code>3</code>)</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>a <code>numeric</code> value adjusting the opacity of colours
representing the correlation statistics (default: <code>1</code>)</p></dd>


<dt id="arg-genebar">genebar<a class="anchor" aria-label="anchor" href="#arg-genebar"></a></dt>
<dd><p>a <code>logical</code> value indicating whether bars representing
the genes should be included in the plot (default: <code>TRUE</code>)</p></dd>


<dt id="arg-genebar-ntracks">genebar_ntracks<a class="anchor" aria-label="anchor" href="#arg-genebar-ntracks"></a></dt>
<dd><p>an <code>integer</code> value indicating the number of
tracks to be included in the gene bar (default: <code>NULL</code>)</p></dd>


<dt id="arg-genebar-label-pos">genebar_label_pos<a class="anchor" aria-label="anchor" href="#arg-genebar-label-pos"></a></dt>
<dd><p>a <code>numeric</code> value indicating the relative
position of gene labels with respect to each gene bar (default: <code>3.6</code>)</p></dd>


<dt id="arg-genebar-label-size">genebar_label_size<a class="anchor" aria-label="anchor" href="#arg-genebar-label-size"></a></dt>
<dd><p>a <code>numeric</code> value defining the size of each
gene label (default: <code>4.25</code>)</p></dd>


<dt id="arg-genebar-line-size">genebar_line_size<a class="anchor" aria-label="anchor" href="#arg-genebar-line-size"></a></dt>
<dd><p>a <code>numeric</code> value defining the line size of
each gene bar (default: <code>0.8</code>)</p></dd>


<dt id="arg-label-size">label_size<a class="anchor" aria-label="anchor" href="#arg-label-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of each label
(default: <code>3.5</code>)</p></dd>


<dt id="arg-highlights">highlights<a class="anchor" aria-label="anchor" href="#arg-highlights"></a></dt>
<dd><p>a <code>character</code> <code>vector</code> defining a set of markers to
highlight in the plot (default: <code>NULL</code>)</p></dd>


<dt id="arg-highlights-cat">highlights_cat<a class="anchor" aria-label="anchor" href="#arg-highlights-cat"></a></dt>
<dd><p>a <code>character</code> <code>vector</code> defining the category for
each highlighted marker (default: <code>NULL</code>)</p></dd>


<dt id="arg-highlights-label">highlights_label<a class="anchor" aria-label="anchor" href="#arg-highlights-label"></a></dt>
<dd><p>a <code>logical</code> value indicating whether highlighted
points should be labelled (default: <code>TRUE</code>)</p></dd>


<dt id="arg-highlights-shape">highlights_shape<a class="anchor" aria-label="anchor" href="#arg-highlights-shape"></a></dt>
<dd><p>a value defining the shape for highlighted points
(default: <code>22</code>)</p></dd>


<dt id="arg-highlights-nolabel-shape">highlights_nolabel_shape<a class="anchor" aria-label="anchor" href="#arg-highlights-nolabel-shape"></a></dt>
<dd><p>a value defining the shape for points
which are not highlighted (default: <code>21</code>)</p></dd>


<dt id="arg-highlights-sort">highlights_sort<a class="anchor" aria-label="anchor" href="#arg-highlights-sort"></a></dt>
<dd><p>a <code>logical</code> value indicating whether to sort highlight
group label levels (default: <code>TRUE</code>)</p></dd>


<dt id="arg-highlights-colours">highlights_colours<a class="anchor" aria-label="anchor" href="#arg-highlights-colours"></a></dt>
<dd><p>a <code>character</code> <code>vector</code> specifying colours for
highlighted points (default: <code>NULL</code>)</p></dd>


<dt id="arg-highlights-title">highlights_title<a class="anchor" aria-label="anchor" href="#arg-highlights-title"></a></dt>
<dd><p>a <code>character</code> string providing a title for the legend
corresponding to the highlighted points (default: <code>"Group"</code>)</p></dd>


<dt id="arg-title">title<a class="anchor" aria-label="anchor" href="#arg-title"></a></dt>
<dd><p>a <code>character</code> string providing a title for the plot
(default: <code>NULL</code>)</p></dd>


<dt id="arg-title-size">title_size<a class="anchor" aria-label="anchor" href="#arg-title-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of the title text
for the plot (default: <code>NULL</code>)</p></dd>


<dt id="arg-title-center">title_center<a class="anchor" aria-label="anchor" href="#arg-title-center"></a></dt>
<dd><p>a <code>logical</code> value indicating whether the plot title
should be centered (default: <code>FALSE</code>)</p></dd>


<dt id="arg-axis-text-size">axis_text_size<a class="anchor" aria-label="anchor" href="#arg-axis-text-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of the axis
text for the plot (default: <code>14</code>)</p></dd>


<dt id="arg-axis-title-size">axis_title_size<a class="anchor" aria-label="anchor" href="#arg-axis-title-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of the axis
title text for the plot (default: <code>16</code>)</p></dd>


<dt id="arg-legend">legend<a class="anchor" aria-label="anchor" href="#arg-legend"></a></dt>
<dd><p>a <code>logical</code>  value indicating whether a legend corresponding
to the displayed groups should be included (default: <code>TRUE</code>)</p></dd>


<dt id="arg-legend-text-size">legend_text_size<a class="anchor" aria-label="anchor" href="#arg-legend-text-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of the legend
text (default: <code>12</code>)</p></dd>


<dt id="arg-legend-title-size">legend_title_size<a class="anchor" aria-label="anchor" href="#arg-legend-title-size"></a></dt>
<dd><p>a <code>numeric</code> value indicating the size of the legend
title (default: <code>12</code>)</p></dd>


<dt id="arg-point-padding">point_padding<a class="anchor" aria-label="anchor" href="#arg-point-padding"></a></dt>
<dd><p>a <code>numeric</code> value indicating the relative distance of
labels from plotted points (default: <code>0</code></p></dd>


<dt id="arg-nudge-x">nudge_x<a class="anchor" aria-label="anchor" href="#arg-nudge-x"></a></dt>
<dd><p>a <code>numeric</code> value indicating the degree to which label
placement on the x-axis should be adjusted (default: <code>0</code>)</p></dd>


<dt id="arg-nudge-y">nudge_y<a class="anchor" aria-label="anchor" href="#arg-nudge-y"></a></dt>
<dd><p>a <code>numeric</code> value indicating the degree to which label
placement on the y-axis should be adjusted (default: <code>0</code>)</p></dd>


<dt id="arg-nudge-y-top">nudge_y_top<a class="anchor" aria-label="anchor" href="#arg-nudge-y-top"></a></dt>
<dd><p>a <code>numeric</code> value indicating the degree to which the top
marker should be adjusted on the y-axis by a proportion of the y-axis limit
(default: <code>0.06</code>)</p></dd>


<dt id="arg-ylim-prob">ylim_prob<a class="anchor" aria-label="anchor" href="#arg-ylim-prob"></a></dt>
<dd><p>a <code>numeric</code> value defining the upper y-axis limit for
probability plots (default: <code>1</code>)</p></dd>


<dt id="arg-assoc-plot-size">assoc_plot_size<a class="anchor" aria-label="anchor" href="#arg-assoc-plot-size"></a></dt>
<dd><p>a <code>numeric</code> value determining the size of the
association plot (default: <code>NULL</code>)</p></dd>


<dt id="arg-genebar-plot-size">genebar_plot_size<a class="anchor" aria-label="anchor" href="#arg-genebar-plot-size"></a></dt>
<dd><p>a <code>numeric</code> value determining the size of the gene
bar plot (default: <code>NULL</code>)</p></dd>


<dt id="arg-legend-plot-dist">legend_plot_dist<a class="anchor" aria-label="anchor" href="#arg-legend-plot-dist"></a></dt>
<dd><p>a <code>numeric</code> value defining the distance and size of
the legend from the bottom of the regional plot (default = <code>NULL</code>)</p></dd>


<dt id="arg-plot-width">plot_width<a class="anchor" aria-label="anchor" href="#arg-plot-width"></a></dt>
<dd><p>a <code>numeric</code> value indicating the width of the plot
(default: <code>9</code>)</p></dd>


<dt id="arg-plot-height">plot_height<a class="anchor" aria-label="anchor" href="#arg-plot-height"></a></dt>
<dd><p>a <code>numeric</code> value indicating the height of the plot
(default: <code>7</code>)</p></dd>


<dt id="arg-girafe">girafe<a class="anchor" aria-label="anchor" href="#arg-girafe"></a></dt>
<dd><p>a <code>logical</code> value indicating whether an interactive plot
should be turned into an interactive graphic using
<a href="https://davidgohel.github.io/ggiraph/reference/girafe.html" class="external-link">girafe()</a> (default = <code>TRUE</code>)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p><code>fig_region</code> returns a regional plot visualising associations
of markers within a genomic region.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>James Staley <a href="mailto:jrstaley95@gmail.com">jrstaley95@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu">fig_region</span><span class="op">(</span></span></span>
<span class="r-in"><span>  data <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="geni_test_region.html">geni_test_region</a></span><span class="op">$</span><span class="va">assoc</span>,</span></span>
<span class="r-in"><span>  corr <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="geni_test_region.html">geni_test_region</a></span><span class="op">$</span><span class="va">corr</span>,</span></span>
<span class="r-in"><span>  build <span class="op">=</span> <span class="fl">37</span>,</span></span>
<span class="r-in"><span>  axis_text_size <span class="op">=</span> <span class="fl">11</span>,</span></span>
<span class="r-in"><span>  axis_title_size <span class="op">=</span> <span class="fl">12</span>,</span></span>
<span class="r-in"><span>  genebar_label_size <span class="op">=</span> <span class="fl">3.5</span>,</span></span>
<span class="r-in"><span>  legend_text_size <span class="op">=</span> <span class="fl">10</span>,</span></span>
<span class="r-in"><span>  legend_title_size <span class="op">=</span> <span class="fl">10</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="fig_region-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Notes:</span></span></span>
<span class="r-in"><span><span class="co">#   (i) corr has to have the same markers as assoc in the same order</span></span></span>
<span class="r-in"><span><span class="co">#   (ii) by default fig_region assumes corr contains correlation</span></span></span>
<span class="r-in"><span><span class="co">#        statistics that have not been squared</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by James Staley, Wes Spiller.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

