<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="fig_region_stack creates a stacked regional association plot."><title>Stacked regional plot — fig_region_stack • geni.plots</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Stacked regional plot — fig_region_stack"><meta property="og:description" content="fig_region_stack creates a stacked regional association plot."><meta property="og:image" content="https://jrs95.github.io/geni.plots/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">geni.plots</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/geni_plots.html">GENI plots</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jrs95/geni.plots/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Stacked regional plot</h1>
      <small class="dont-index">Source: <a href="https://github.com/jrs95/geni.plots/blob/HEAD/R/fig_region.R" class="external-link"><code>R/fig_region.R</code></a></small>
      <div class="d-none name"><code>fig_region_stack.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>fig_region_stack</code> creates a stacked regional association plot.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fig_region_stack</span><span class="op">(</span></span>
<span>  <span class="va">data</span>,</span>
<span>  <span class="va">traits</span>,</span>
<span>  corr <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  corr_top <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top_marker <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  r2 <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  build <span class="op">=</span> <span class="fl">38</span>,</span>
<span>  prob <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  thresh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  thresh_colour <span class="op">=</span> <span class="st">"grey50"</span>,</span>
<span>  x_min <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  x_max <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  y_title <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  point_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  genebar <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  genebar_ntracks <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genebar_label_pos <span class="op">=</span> <span class="fl">3.6</span>,</span>
<span>  genebar_label_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  genebar_line_size <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">3.5</span>,</span>
<span>  highlights <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlights_cat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlights_label <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  highlights_shape <span class="op">=</span> <span class="fl">22</span>,</span>
<span>  highlights_nolabel_shape <span class="op">=</span> <span class="fl">21</span>,</span>
<span>  highlights_sort <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  highlights_colours <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  highlights_title <span class="op">=</span> <span class="st">"Group"</span>,</span>
<span>  title_size <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  title_center <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  axis_text_size <span class="op">=</span> <span class="fl">14</span>,</span>
<span>  axis_title_size <span class="op">=</span> <span class="fl">16</span>,</span>
<span>  legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  legend_text_size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  legend_title_size <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  point_padding <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  nudge_x <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  nudge_y <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  nudge_y_top <span class="op">=</span> <span class="fl">0.06</span>,</span>
<span>  ylim_prob <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  assoc_plot_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  genebar_plot_size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_plot_dist <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_width <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  plot_height <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  girafe <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>data</dt>
<dd><p>a <code>data.frame</code> containing the association statistics for each
marker within a genomic region with the following columns:</p><ul><li><p><code>marker</code> 
the genomic marker identifier (e.g. rsID)</p></li>
<li><p><code>chr</code> 
the chromosome for each genomic marker</p></li>
<li><p><code>pos</code> 
the genomic position for each genomic marker

and one of the following:</p></li>
<li><p><code>pvalue_1</code>, <code>pvalue_2</code>, <code>pvalue_3</code>, etc. 
the association p-values for each genomic marker and trait</p></li>
<li><p><code>z_1</code>, <code>z_2</code>, <code>z_3</code>, etc. 
the association z-statistics for each genomic marker and trait</p></li>
<li><p><code>prob_1</code>, <code>prob_2</code>, <code>prob_3</code>, etc. 
the association probabilities for each genomic marker and trait</p></li>
</ul></dd>


<dt>traits</dt>
<dd><p>a <code>character</code> <code>vector</code> of trait names</p></dd>


<dt>corr</dt>
<dd><p>a <code>numeric</code> <code>matrix</code> of correlation statistics between
the markers (default: <code>NULL</code>)</p></dd>


<dt>corr_top</dt>
<dd><p>a <code>numeric</code> <code>vector</code> of correlation statistics between the
top marker and the rest of the markers (default: <code>NULL</code>)</p></dd>


<dt>top_marker</dt>
<dd><p>a <code>character</code> value depicting the marker to plot the
correlation statistics of the rest of the markers against (default: <code>NULL</code>)</p></dd>


<dt>r2</dt>
<dd><p>a <code>logical</code> value indicating whether the set of correlation
statistics entered in <code>corr</code> or <code>corr_top</code> are squared (default: <code>FALSE</code>)</p></dd>


<dt>build</dt>
<dd><p>a <code>numeric</code> value indicating the genome build used to determine
genomic position (default: <code>38</code> representing human assembly GRCh38)</p></dd>


<dt>prob</dt>
<dd><p>a <code>logical</code> value indicating whether probability statistics
should be plotted instead of -log10(p-values) (default: <code>FALSE</code>)</p></dd>


<dt>interactive</dt>
<dd><p>a <code>logical</code> value indicating whether the plot
should be interactive (default: <code>FALSE</code>)</p></dd>


<dt>thresh</dt>
<dd><p>a <code>numeric</code> <code>vector</code> providing the p-value thresholds to be
plotted (default: <code>NULL</code>)</p></dd>


<dt>thresh_colour</dt>
<dd><p>a <code>character</code> <code>vector</code> indicating the colours of the
lines indicating the p-value thresholds (default: <code>"grey50"</code>)</p></dd>


<dt>x_min</dt>
<dd><p>a <code>numeric</code> value depicting the minimum plotted x-axis value
representing the start of the genomic region (default: <code>NULL</code>)</p></dd>


<dt>x_max</dt>
<dd><p>a <code>numeric</code> value depicting the maximum plotted x-axis value
representing the end of the genomic region (default: <code>NULL</code>)</p></dd>


<dt>y_title</dt>
<dd><p>a <code>character</code> string defining the title of the y-axis
(default: <code>NULL</code>)</p></dd>


<dt>point_size</dt>
<dd><p>a <code>numeric</code> value indicating the size of each point
(default: <code>3</code>)</p></dd>


<dt>alpha</dt>
<dd><p>a <code>numeric</code> value adjusting the opacity of colours
representing the correlation statistics (default: <code>1</code>)</p></dd>


<dt>genebar</dt>
<dd><p>a <code>logical</code> value indicating whether bars representing
the genes should be included in the plot (default: <code>TRUE</code>)</p></dd>


<dt>genebar_ntracks</dt>
<dd><p>an <code>integer</code> value indicating the number of
tracks to be included in the gene bar (default: <code>NULL</code>)</p></dd>


<dt>genebar_label_pos</dt>
<dd><p>a <code>numeric</code> value indicating the relative
position of gene labels with respect to each gene bar (default: <code>3.6</code>)</p></dd>


<dt>genebar_label_size</dt>
<dd><p>a <code>numeric</code> value defining the size of each
gene label (default: <code>4.25</code>)</p></dd>


<dt>genebar_line_size</dt>
<dd><p>a <code>numeric</code> value defining the line size of
each gene bar (default: <code>0.8</code>)</p></dd>


<dt>label_size</dt>
<dd><p>a <code>numeric</code> value indicating the size of each label
(default: <code>3.5</code>)</p></dd>


<dt>highlights</dt>
<dd><p>a <code>character</code> <code>vector</code> defining a set of markers to
highlight in the plot (default: <code>NULL</code>)</p></dd>


<dt>highlights_cat</dt>
<dd><p>a <code>character</code> <code>vector</code> defining the category for
each highlighted marker (default: <code>NULL</code>)</p></dd>


<dt>highlights_label</dt>
<dd><p>a <code>logical</code> value indicating whether highlighted
points should be labelled (default: <code>TRUE</code>)</p></dd>


<dt>highlights_shape</dt>
<dd><p>a value defining the shape for highlighted points
(default: <code>22</code>)</p></dd>


<dt>highlights_nolabel_shape</dt>
<dd><p>a value defining the shape for points
which are not highlighted (default: <code>21</code>)</p></dd>


<dt>highlights_sort</dt>
<dd><p>a <code>logical</code> value indicating whether to sort highlight
group label levels (default: <code>TRUE</code>)</p></dd>


<dt>highlights_colours</dt>
<dd><p>a <code>character</code> <code>vector</code> specifying colours for
highlighted points (default: <code>NULL</code>)</p></dd>


<dt>highlights_title</dt>
<dd><p>a <code>character</code> string providing a title for the legend
corresponding to the highlighted points (default: <code>"Group"</code>)</p></dd>


<dt>title_size</dt>
<dd><p>a <code>numeric</code> value indicating the size of the title text
for the plot (default: <code>NULL</code>)</p></dd>


<dt>title_center</dt>
<dd><p>a <code>logical</code> value indicating whether the plot title
should be centered (default: <code>FALSE</code>)</p></dd>


<dt>axis_text_size</dt>
<dd><p>a <code>numeric</code> value indicating the size of the axis
text for the plot (default: <code>14</code>)</p></dd>


<dt>axis_title_size</dt>
<dd><p>a <code>numeric</code> value indicating the size of the axis
title text for the plot (default: <code>16</code>)</p></dd>


<dt>legend</dt>
<dd><p>a <code>logical</code>  value indicating whether a legend corresponding
to the displayed groups should be included (default: <code>TRUE</code>)</p></dd>


<dt>legend_text_size</dt>
<dd><p>a <code>numeric</code> value indicating the size of the legend
text (default: <code>NULL</code>)</p></dd>


<dt>legend_title_size</dt>
<dd><p>a <code>numeric</code> value indicating the size of the legend
title (default: <code>NULL</code>)</p></dd>


<dt>point_padding</dt>
<dd><p>a <code>numeric</code> value indicating the relative distance of
labels from plotted points (default: <code>0</code></p></dd>


<dt>nudge_x</dt>
<dd><p>a <code>numeric</code> value indicating the degree to which label
placement on the x-axis should be adjusted (default: <code>0</code>)</p></dd>


<dt>nudge_y</dt>
<dd><p>a <code>numeric</code> value indicating the degree to which label
placement on the y-axis should be adjusted (default: <code>0</code>)</p></dd>


<dt>nudge_y_top</dt>
<dd><p>a <code>numeric</code> value indicating the degree to which the top
marker should be adjusted on the y-axis by a proportion of the y-axis limit
(default: <code>0.06</code>)</p></dd>


<dt>ylim_prob</dt>
<dd><p>a <code>numeric</code> value defining the upper y-axis limit for
probability plots (default: <code>1</code>)</p></dd>


<dt>assoc_plot_size</dt>
<dd><p>a <code>numeric</code> value determining the size of the
association plot (default: <code>NULL</code>)</p></dd>


<dt>genebar_plot_size</dt>
<dd><p>a <code>numeric</code> value determining the size of the gene
bar plot (default: <code>NULL</code>)</p></dd>


<dt>legend_plot_dist</dt>
<dd><p>a <code>numeric</code> value defining the distance and size of
the legend from the bottom of the regional plot (default = <code>NULL</code>)</p></dd>


<dt>plot_width</dt>
<dd><p>a <code>numeric</code> value indicating the width of the plot
(default: <code>NULL</code>)</p></dd>


<dt>plot_height</dt>
<dd><p>a <code>numeric</code> value indicating the height of the plot
(default: <code>NULL</code>)</p></dd>


<dt>girafe</dt>
<dd><p>a <code>logical</code> value indicating whether an interactive plot
should be turned into an interactive graphic using
<a href="https://davidgohel.github.io/ggiraph/reference/girafe.html" class="external-link">girafe()</a> (default = <code>TRUE</code>)</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p><code>fig_region_stack</code> returns a stacked regional plot.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>James Staley <a href="mailto:jrstaley95@gmail.com">jrstaley95@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">fig</span> <span class="op">&lt;-</span> <span class="fu">fig_region_stack</span><span class="op">(</span></span></span>
<span class="r-in"><span>  data <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="geni_test_stack_region.html">geni_test_stack_region</a></span><span class="op">$</span><span class="va">assoc</span>,</span></span>
<span class="r-in"><span>  traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Interleukin-6 levels"</span>, <span class="st">"Interleukin-6 receptor levels"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  corr <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="geni_test_stack_region.html">geni_test_stack_region</a></span><span class="op">$</span><span class="va">corr</span>,</span></span>
<span class="r-in"><span>  build <span class="op">=</span> <span class="fl">37</span>,</span></span>
<span class="r-in"><span>  highlights <span class="op">=</span> <span class="st">"rs11265611"</span>,</span></span>
<span class="r-in"><span>  title_center <span class="op">=</span> <span class="cn">TRUE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by James Staley, Wes Spiller.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

