<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="geni.plots">
<title>GENI plots • geni.plots</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="GENI plots">
<meta property="og:description" content="geni.plots">
<meta property="og:image" content="https://jrs95.github.io/geni.plots/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">geni.plots</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.0.4</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/geni_plots.html">GENI plots</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/jrs95/geni.plots/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>GENI plots</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/jrs95/geni.plots/blob/HEAD/vignettes/geni_plots.Rmd" class="external-link"><code>vignettes/geni_plots.Rmd</code></a></small>
      <div class="d-none name"><code>geni_plots.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>The <code>geni.plots</code> R package is a tool for visualising
results from genome-wide association studies (GWAS) and phenome-wide
association studies (PheWAS). The package can currently produce
Manhattan plots, quantile-quantile (QQ) plots, and regional plots using
GWAS summary statistics using the <code><a href="../reference/fig_manhattan.html">fig_manhattan()</a></code>,
<code><a href="../reference/fig_qq.html">fig_qq()</a></code>, and <code><a href="../reference/fig_region.html">fig_region()</a></code> functions,
respectively. In addition, PheWAS plots can be constructed using the
<code><a href="../reference/fig_phewas.html">fig_phewas()</a></code> function.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jrs95/geni.plots" class="external-link">geni.plots</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="functions">Functions<a class="anchor" aria-label="anchor" href="#functions"></a>
</h2>
<p>The package contains the following functions:</p>
<ul>
<li>
<code>fig_manhattan</code>: creates a Manhattan plot for genomic
markers from across the genome, e.g. results from genome-wide
association studies.<br>
</li>
<li>
<code>fig_phewas</code>: creates a plot visualising results from
phenome-wide association studies (PheWAS).<br>
</li>
<li>
<code>fig_qq</code>: creates a quantile-quantile (QQ) plot.<br>
</li>
<li>
<code>fig_region</code>: creates a regional plot, i.e. a scatter
graph of genomic markers associations (e.g. log10(p-values)) with a gene
bar underneath.<br>
</li>
<li>
<code>fig_stack_region</code>: creates a stacked regional
association plot.</li>
</ul>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>The package contains formatted test data downloaded from FinnGen r9
(<a href="https://r9.finngen.fi/variant/1:154454494-A-C" class="external-link">1:154426970-A-C</a>)
and from GWAS studies on IL6 (<a href="https://www.ebi.ac.uk/gwas/studies/GCST90274815" class="external-link">GCST90274815</a>)
and IL6R (<a href="https://www.ebi.ac.uk/gwas/studies/GCST90274815" class="external-link">GCST90274815</a>)
levels from the GWAS Catalog server.</p>
<ul>
<li>
<code>geni_test_manhattan</code>- a <code>data.frame</code> of
results from a GWAS of interleukin-6 levels with <em>p</em>-values &lt;
<code>1e-3</code> (<a href="https://www.ebi.ac.uk/gwas/studies/GCST90274815" class="external-link">GCST90274815</a>).
The genomic coordinates used are GRCh37. A description of this data can
be viewed using <code>help(geni_test_gwas)</code>.<br>
</li>
<li>
<code>geni_test_phewas</code>- a <code>data.frame</code> of PheWAS
results for rs from FinnGen r9 (<a href="https://r9.finngen.fi/variant/1:154454494-A-C" class="external-link">1:154426970-A-C</a>).
A description of this data can be viewed using
<code><a href="../reference/geni_test_phewas.html">help(geni_test_phewas)</a></code>.<br>
</li>
<li>
<code>geni_test_region</code>- a <code>list</code> containing:
<ul>
<li>
<code>assoc</code> - a <code>data.frame</code> of results from a
GWAS of interleukin-6 levels for the GRCh37 region 1:154301970-154551970
(<a href="https://www.ebi.ac.uk/gwas/studies/GCST90274815" class="external-link">GCST90274815</a>).
The genomic coordinates used are GRCh37. A description of this data can
be viewed using <code><a href="../reference/geni_test_region.html">help(geni_test_region)</a></code>.<br>
</li>
<li>
<code>corr</code> - a <code>matrix</code> of correlation statistics
for the markers in <code>assoc</code> using the European samples from <a href="https://www.internationalgenome.org/" class="external-link">1000 Genomes Phase 3</a>.
The genomic coordinates used are GRCh37.<br>
</li>
</ul>
</li>
<li>
<code>geni_test_stack_region</code>- a <code>list</code> containing:
<ul>
<li>
<code>assoc</code> - a <code>data.frame</code> of results from GWAS
studies on interleukin-6 and interleukin 6 receptor levels for the
GRCh37 region 1:154301970-154551970 (<a href="https://www.ebi.ac.uk/gwas/studies/GCST90274815" class="external-link">GCST90274815</a>
and <a href="https://www.ebi.ac.uk/gwas/studies/GCST90088597" class="external-link">GCST90088597</a>).
The genomic coordinates used are GRCh37. A description of this data can
be viewed using <code><a href="../reference/geni_test_region.html">help(geni_test_region)</a></code>.<br>
</li>
<li>
<code>corr</code> - a <code>matrix</code> of correlation statistics
for the markers in <code>assoc</code> using the European samples from <a href="https://www.internationalgenome.org/" class="external-link">1000 Genomes Phase 3</a>.
The genomic coordinates used are GRCh37.</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 class="tabset tabset-fade" id="plots">Plots<a class="anchor" aria-label="anchor" href="#plots"></a>
</h2>
<div class="section level3">
<h3 id="manhattan-plot">Manhattan plot<a class="anchor" aria-label="anchor" href="#manhattan-plot"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_manhattan.html">geni_test_manhattan</a></span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 6</span></span></span>
<span><span class="co">#&gt;   chr       pos    pvalue highlight highlight_shape label</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 1      <span style="text-decoration: underline;">963</span>661 0.000<span style="text-decoration: underline;">154</span>          0               0 <span style="color: #949494;">""</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 1      <span style="text-decoration: underline;">968</span>251 0.000<span style="text-decoration: underline;">538</span>          0               0 <span style="color: #949494;">""</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 1     1<span style="text-decoration: underline;">010</span>040 0.000<span style="text-decoration: underline;">167</span>          0               0 <span style="color: #949494;">""</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 1     1<span style="text-decoration: underline;">112</span>698 0.000<span style="text-decoration: underline;">682</span>          0               0 <span style="color: #949494;">""</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 1     2<span style="text-decoration: underline;">031</span>103 0.000<span style="text-decoration: underline;">089</span>7         0               0 <span style="color: #949494;">""</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 1     2<span style="text-decoration: underline;">036</span>650 0.000<span style="text-decoration: underline;">207</span>          0               0 <span style="color: #949494;">""</span></span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fig_manhattan.html">fig_manhattan</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_manhattan.html">geni_test_manhattan</a></span>,</span>
<span>  block_thresh <span class="op">=</span> <span class="fl">1e-4</span>,</span>
<span>  label_box <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="geni_plots_files/figure-html/manhattan-1.png" width="100%"></p>
<p>Further details regarding functionality of the
<code>fig_manhattan</code> function can be viewed using the command
<code><a href="../reference/fig_manhattan.html">help(fig_manhattan)</a></code>.</p>
</div>
<div class="section level3">
<h3 id="phewas-plot">PheWAS plot<a class="anchor" aria-label="anchor" href="#phewas-plot"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_phewas.html">geni_test_phewas</a></span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt;   pvalue  sign group                    label                              text </span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                              <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> 0.375     -<span style="color: #BB0000;">1</span> AB1: Infectious diseases Actinomycosis                      Phen…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> 0.050<span style="text-decoration: underline;">6</span>     1 AB1: Infectious diseases Amoebiasis                         Phen…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> 0.112      1 AB1: Infectious diseases Anogenital herpesviral [herpes si… Phen…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> 0.287     -<span style="color: #BB0000;">1</span> AB1: Infectious diseases Aspergillosis                      Phen…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> 0.035<span style="text-decoration: underline;">2</span>    -<span style="color: #BB0000;">1</span> AB1: Infectious diseases Atypical virus infections of cent… Phen…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> 0.011<span style="text-decoration: underline;">3</span>     1 AB1: Infectious diseases Bacterial, viral and other infect… Phen…</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fig_phewas.html">fig_phewas</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_phewas.html">geni_test_phewas</a></span>,</span>
<span>  axis_text_angle <span class="op">=</span> <span class="op">-</span><span class="fl">70</span>,</span>
<span>  axis_text_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">3.5</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="geni_plots_files/figure-html/phewas-1.png" width="100%"></p>
<p>Note: this figure can be made interactive by setting
<code>interactive = TRUE</code>.</p>
<p>Further details regarding functionality of the
<code>fig_phewas</code> function can be viewed using the command
<code><a href="../reference/fig_phewas.html">help(fig_phewas)</a></code>.</p>
</div>
<div class="section level3">
<h3 class="tabset tabset-fade tabset-pills" id="qq-plot">QQ plot<a class="anchor" aria-label="anchor" href="#qq-plot"></a>
</h3>
<div class="section level4">
<h4 id="simple">Simple<a class="anchor" aria-label="anchor" href="#simple"></a>
</h4>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fig_qq.html">fig_qq</a></span><span class="op">(</span></span>
<span>  pvalues <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_phewas.html">geni_test_phewas</a></span><span class="op">$</span><span class="va">pvalue</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="geni_plots_files/figure-html/qq-1.png" width="100%"></p>
<p>Further details regarding functionality of the <code>fig_qq</code>
function can be viewed using the command <code><a href="../reference/fig_qq.html">help(fig_qq)</a></code>.</p>
</div>
<div class="section level4">
<h4 id="categories">Categories<a class="anchor" aria-label="anchor" href="#categories"></a>
</h4>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fig_qq.html">fig_qq</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_phewas.html">geni_test_phewas</a></span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">pvalue</span>, <span class="va">group</span>, <span class="va">label</span>, <span class="va">text</span><span class="op">)</span>,</span>
<span>  legend_title <span class="op">=</span> <span class="st">"Category"</span>,</span>
<span>  legend_title_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  legend_text_size <span class="op">=</span> <span class="fl">8</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="geni_plots_files/figure-html/qq-cats-1.png" width="100%"></p>
<p>Note: this figure can be made interactive by setting
<code>interactive = TRUE</code>.</p>
<p>Further details regarding functionality of the <code>fig_qq</code>
function can be viewed using the command <code><a href="../reference/fig_qq.html">help(fig_qq)</a></code>.</p>
</div>
</div>
<div class="section level3">
<h3 class="tabset tabset-fade tabset-pills" id="regional-plot">Regional plot<a class="anchor" aria-label="anchor" href="#regional-plot"></a>
</h3>
<div class="section level4">
<h4 id="regional-plot-1">Regional plot<a class="anchor" aria-label="anchor" href="#regional-plot-1"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fig_region.html">fig_region</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_region.html">geni_test_region</a></span><span class="op">$</span><span class="va">assoc</span>,</span>
<span>  corr <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_region.html">geni_test_region</a></span><span class="op">$</span><span class="va">corr</span>,</span>
<span>  build <span class="op">=</span> <span class="fl">37</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="geni_plots_files/figure-html/regional-1.png" width="100%"></p>
<p>Note:</p>
<ol style="list-style-type: decimal">
<li>this figure can be made interactive by setting
<code>interactive = TRUE</code><br>
</li>
<li>
<code>corr</code> has to have the same markers as <code>assoc</code>
in the same order<br>
</li>
<li>by default <code>fig_region</code> assumes <code>corr</code> (or
<code>corr_top</code>) contains correlation statistics that have not
been squared, if <code>corr</code> is a squared correlation matrix set
<code>r2 = TRUE</code>
</li>
</ol>
<p>Further details regarding functionality of the
<code>fig_region</code> function can be viewed using the command
<code><a href="../reference/fig_region.html">help(fig_region)</a></code>.</p>
</div>
<div class="section level4">
<h4 id="stacked-regional-plot">Stacked regional plot<a class="anchor" aria-label="anchor" href="#stacked-regional-plot"></a>
</h4>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/fig_region_stack.html">fig_region_stack</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_stack_region.html">geni_test_stack_region</a></span><span class="op">$</span><span class="va">assoc</span>,</span>
<span>  traits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Interleukin-6 levels"</span>, <span class="st">"Interleukin-6 receptor levels"</span><span class="op">)</span>,</span>
<span>  corr <span class="op">=</span> <span class="fu">geni.plots</span><span class="fu">::</span><span class="va"><a href="../reference/geni_test_stack_region.html">geni_test_stack_region</a></span><span class="op">$</span><span class="va">corr</span>,</span>
<span>  build <span class="op">=</span> <span class="fl">37</span>,</span>
<span>  highlights <span class="op">=</span> <span class="st">"rs11265611"</span>,</span>
<span>  title_center <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="geni_plots_files/figure-html/regional-stack-1.png" width="100%"></p>
<p>Note:</p>
<ol style="list-style-type: decimal">
<li>this figure can be made interactive by setting
<code>interactive = TRUE</code><br>
</li>
<li>
<code>corr</code> has to have the same markers as <code>assoc</code>
in the same order<br>
</li>
<li>by default <code>fig_region_stack</code> assumes <code>corr</code>
(or <code>corr_top</code>) contains correlation statistics that have not
been squared, if <code>corr</code> is a squared correlation matrix set
<code>r2 = TRUE</code>
</li>
</ol>
<p>Further details regarding functionality of the
<code>fig_region_stack</code> function can be viewed using the command
<code><a href="../reference/fig_region_stack.html">help(fig_region_stack)</a></code>.</p>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by James Staley, Wes Spiller.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
